### ISSUE TITLE
[no-unused-vars] false positive on "?." operator

### ISSUE THREAD

**Issue by higherorderfunctor** [typescript-eslint/typescript-eslint#2525]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```
{
  root: true,
  extends: [
    'airbnb-typescript/base',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended',
    'plugin:promise/recommended',
  ],
  parserOptions: {
    project: './tsconfig.json',
  },
  plugins: [
    'simple-import-sort',
    'promise',
    'eslint-plugin-tsdoc'
   ],
  rules: {
    'simple-import-sort/sort': 'error',
    'no-underscore-dangle': [
      'error',
      { allow: ['_id', '_read', '_write', '_final'] },
    ],
    'no-warning-comments': [
      'warn', {
        terms: ["todo", "fixme"],
        location: 'start'
      }
    ],
    'tsdoc/syntax': 'error'
  },
  'reportUnusedDisableDirectives': true,
}
```

```TS
interface A {
  a?: string;
}

export function testA(opts: A): void {
  if (opts?.a) { // [eslint/Warning/@typescript-eslint/no-unused-vars] 'opts' is defined but never used.
    // noop
  }
}

interface B {
  b?: string;
}

export function testB(opts: B): void {
  if (opts.b) { // work okay
    // noop
  }
}
```

**Expected Result**

`testA` should pass linting like `testB`.

**Actual Result**

`testA` reports `opts` as unused when it is.

**Additional Info**

N/A

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `12.18.3` |

**Comment by bradzacher** [comment]
As per #2473
You likely have two versions of ESLint / the tooling installed.

If you are sure you don't, please create a repro repo and we can reopen and investigate.

**Comment by higherorderfunctor** [comment]
@bradzacher thank you for the feedback.  Note sure the ultimate cause, as I was not able to reproduce in a new repo, but removing my node_modules folder and reinstalling all the dependencies finally fixed it.

**Comment by laozhu** [comment]
Maybe I have the some problem, always return false with optional-chaining

**1. with optional-chaining**

![image](https://user-images.githubusercontent.com/1212457/93321177-7bff1280-f844-11ea-85b7-7ccb52d3f0d5.png)

**2. without optional-chaining**

![image](https://user-images.githubusercontent.com/1212457/93321618-f2037980-f844-11ea-873f-124a7c39f7a7.png)
